### PR TITLE
Do not serialize data type if not initialized

### PR DESCRIPTION
**Pull request by tanjialiang**
QueryDataWriter::writeSummary() crashes out of nullptr due to empty writes not initializing data type. This PR fixes it by not serializing data type if not initialized.

### CODE REVIEW

File: velox/exec/QueryDataWriter.cpp [hunk 1]
```diff
@@ -42,6 +42,11 @@ void QueryDataWriter::write(const RowVectorPtr& rows) {
   if (FOLLY_UNLIKELY(finished_)) {
     return;
   }
+  if (dataType_ == nullptr) {
+    dataType_ = rows->type();
+  } else {
+    VELOX_CHECK(dataType_->kindEquals(rows->type()));
```
Comment (duanmeng) [inline velox/exec/QueryDataWriter.cpp:48]: Should we remove this check and use `FOLLY_UNLIKELY(dataType_ == nullptr)` as this is a performance critical path inside `Operator::addInput`?

**Comment by duanmeng** [comment]
Could we remove `dataType_` from `QueryDataWriter` and the summary file? The query replayed framework now extracts input data type from the plan node see `QueryTraceUtil::getDataType`. Also, we can record the number of batches recorded by the `QueryDataWriter`, and write to the summary file. What do you think? cc @xiaoxmeng

**Comment by tanjialiang** [comment]
> @tanjialiang Thanks for your fix. Could we remove `dataType_` from `QueryDataWriter` and the summary file? The query replayed framework now extracts input data type from the plan node see `QueryTraceUtil::getDataType`. Also, we can record the number of batches recorded by the `QueryDataWriter`, and write to the summary file. What do you think? cc @xiaoxmeng

It's actually good to be able to view that in the summary file. I used it multiple times to view the data type while debugging.

File: velox/exec/QueryDataWriter.cpp (hunk 1 above)
Comment (tanjialiang) [reply velox/exec/QueryDataWriter.cpp:48]: It's actually not that bad as we call it once per batch. But yeah sure we can remove it and have what you mentioned above

**Comment by duanmeng** [comment]
> > @tanjialiang Thanks for your fix. Could we remove `dataType_` from `QueryDataWriter` and the summary file? The query replayed framework now extracts input data type from the plan node see `QueryTraceUtil::getDataType`. Also, we can record the number of batches recorded by the `QueryDataWriter`, and write to the summary file. What do you think? cc @xiaoxmeng
> 
> It's actually good to be able to view that in the summary file. I used it multiple times to view the data type while debugging.

Is there any chance to use `QueryTraceUtil::getDataType` when write summary files then we can remove the data type capture logic from `QueryDataWriter::write` along with those `ifs` and checks. We can extract data type during write summary, it is only called at the end of the data tracing, it is not a performance critical path.

**Comment by tanjialiang** [comment]
> > > @tanjialiang Thanks for your fix. Could we remove `dataType_` from `QueryDataWriter` and the summary file? The query replayed framework now extracts input data type from the plan node see `QueryTraceUtil::getDataType`. Also, we can record the number of batches recorded by the `QueryDataWriter`, and write to the summary file. What do you think? cc @xiaoxmeng
> > 
> > 
> > It's actually good to be able to view that in the summary file. I used it multiple times to view the data type while debugging.
> 
> Is there any chance to use `QueryTraceUtil::getDataType` when write summary files then we can remove the data type capture logic from `QueryDataWriter::write` along with those `ifs` and checks. We can extract data type during write summary, it is only called at the end of the data tracing, it is not a performance critical path.

That sounds like a good idea. Let's do that in a separate refactor? Need to get these fix changes in quickly to use for debugging.

**Comment by duanmeng** [comment]
> > > > @tanjialiang Thanks for your fix. Could we remove `dataType_` from `QueryDataWriter` and the summary file? The query replayed framework now extracts input data type from the plan node see `QueryTraceUtil::getDataType`. Also, we can record the number of batches recorded by the `QueryDataWriter`, and write to the summary file. What do you think? cc @xiaoxmeng
> > > 
> > > 
> > > It's actually good to be able to view that in the summary file. I used it multiple times to view the data type while debugging.
> > 
> > 
> > Is there any chance to use `QueryTraceUtil::getDataType` when write summary files then we can remove the data type capture logic from `QueryDataWriter::write` along with those `ifs` and checks. We can extract data type during write summary, it is only called at the end of the data tracing, it is not a performance critical path.
> 
> That sounds like a good idea. Let's do that in a separate refactor? Need to get these fix changes in quickly to use for debugging.

Sure

**Review by duanmeng — APPROVED** [review]
@tanjialiang LGTM thanks for your fix.

**Review by xiaoxmeng — APPROVED** [review]
@tanjialiang thanks!

**Pull request closed without merging**